File: LeetCode/0126.Word-Ladder-II.cpp
// 126. Word Ladder II

class Solution {
    unordered_map<string,int> mp;
    vector<vector<string>> ans;
    string begin;

private:
    void dfs(string word, vector<string> &seq) {
        if(word == begin) {
            reverse(seq.begin(),seq.end());
            ans.push_back(seq);
            reverse(seq.begin(),seq.end());
            return;
        }

        int cnt = word.size(); 
        int steps = mp[word];
        for(int i=0; i<cnt; i++) {
            char original = word[i];
            for(char ch='a'; ch<='z'; ch++) {
                word[i] = ch;
                if(mp.find(word) != mp.end() && mp[word]+1==steps) {
                    seq.push_back(word);
                    dfs(word, seq);
                    seq.pop_back();
                }
            }
            word[i] = original;
        }
    }

public:
    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        begin = beginWord;
        queue<string> q;
        q.push(beginWord);
        unordered_set<string> s(wordList.begin(),wordList.end());
        s.erase(beginWord);
        mp[beginWord] = 1;
        int cnt = beginWord.size();

        while(!q.empty()) {
            string word = q.front();
            int steps = mp[word];
            q.pop();
            if(word == endWord) break;
            for(int i=0; i<cnt; i++) {
                char original = word[i];
                for(char ch='a'; ch<='z'; ch++) {
                    word[i] = ch;
                    if(s.find(word) != s.end()) {
                        q.push(word);
                        mp[word] = steps+1;
                        s.erase(word);
                    }
                }
                word[i] = original;
            }
        }
        if(mp.find(endWord) != mp.end()) {
            vector<string> seq;
            seq.push_back(endWord);
            dfs(endWord, seq);
        }
        return ans;
    }
};

File: LeetCode/0126-Word-Ladder-II.java
// 126. Word Ladder II

class Solution {
    Map<String,List<String>> parent;

    public void dfs(String currWord, String beginWord, List<String> path, List<List<String>> ans) {
        if (currWord.equals(beginWord)) {
            List<String> tmp = new ArrayList<>(path);
            Collections.reverse(tmp);
            ans.add(new ArrayList<>(tmp));
            return;
        }

        for (String prev : parent.get(currWord)) {
            path.add(prev);
            dfs(prev, beginWord, path, ans);
            path.remove(path.size() - 1);
        }
    }

    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        Set<String> st = new HashSet<>(wordList);
        List<List<String>> ans = new ArrayList<>();
        parent = new HashMap<>();

        if (!st.contains(endWord)) return ans;

        Queue<String> q = new LinkedList<>();
        q.add(beginWord);
        st.remove(beginWord);

        while (!q.isEmpty()) {
            int sz = q.size();
            Set<String> removeStr = new HashSet<>();
            while (sz-- > 0) {
                String curr = q.poll();
                char[] arr = curr.toCharArray();
                for (int i = 0; i < arr.length; i++) {
                    char org = arr[i];
                    for (char ch = 'a'; ch <= 'z'; ch++) {
                        arr[i] = ch;
                        String nextStr = new String(arr);
                        if (!st.contains(nextStr)) continue;
                        if (!removeStr.contains(nextStr)) {
                            q.add(nextStr);
                            removeStr.add(nextStr);
                        }
                        parent.computeIfAbsent(nextStr, k -> new ArrayList<>()).add(curr);
                    }
                    arr[i] = org;
                }
            }
            st.removeAll(removeStr);
        }

        if (!parent.containsKey(endWord)) return ans;

        List<String> path = new ArrayList<>(Arrays.asList(endWord));
        dfs(endWord, beginWord, path, ans);
    
        return ans;
    }
}
